File: simple_poker_hands_java/src/main/java/local/kata/simplepokerhands/Solver.java
package local.kata.simplepokerhands;

import java.util.Comparator;
import java.util.List;
import java.util.Optional;
import java.util.stream.Stream;
import local.kata.simplepokerhands.cards.Card;
import local.kata.simplepokerhands.cards.CardStrength;
import local.kata.simplepokerhands.cards.CardSuit;
import local.kata.simplepokerhands.poker.*;
import local.kata.simplepokerhands.tuples.Pair;
import local.kata.simplepokerhands.tuples.Quadruple;
import local.kata.simplepokerhands.tuples.Quintuple;
import local.kata.simplepokerhands.tuples.Triple;

public final class Solver {
    private final List<Card> cards;

    public Solver(List<Card> cards) {
        this.cards = cards;
    }

    public Optional<? extends PokerHand> solve() {
        return Stream.<Optional<? extends PokerHand>>of(
                        findStraightFlush(cards),
                        findFourOfAKind(cards),
                        findFullHouse(cards),
                        findFlush(cards),
                        findStraight(cards),
                        findThreeOfAKind(cards),
                        findTwoPair(cards),
                        findOnePair(cards),
                        findHighCard(cards))
                .flatMap(Optional::stream)
                .findFirst();
    }

    public static Optional<HighCard> findHighCard(List<Card> cards) {
        return cards.stream()
                .max(Comparator.comparingInt(item -> item.strength().ordinal()))
                .map(HighCard::new);
    }

    public static Optional<OnePair> findOnePair(List<Card> cards) {
        Optional<List<Card>> pairOpt = findSameCards(cards, 2);
        if (pairOpt.isEmpty()) return Optional.empty();
        List<Card> pair = pairOpt.get();

        List<Card> list = cards.stream()
                .filter(card -> card.strength() != pair.getFirst().strength())
                .toList();
        Optional<HighCard> kickerOpt = findHighCard(list);
        if (kickerOpt.isEmpty()) return Optional.empty();
        HighCard kicker = kickerOpt.get();

        return Optional.of(new OnePair(new Pair<>(pair.get(0), pair.get(1)), kicker.card));
    }

    public static Optional<TwoPair> findTwoPair(List<Card> cards) {
        Optional<OnePair> firstOpt = findOnePair(cards);
        if (firstOpt.isEmpty()) return Optional.empty();
        OnePair first = firstOpt.get();

        List<Card> list = cards.stream()
                .filter(card -> card.strength() != first.cards.first().strength())
                .toList();
        Optional<OnePair> secondOpt = findOnePair(list);
        if (secondOpt.isEmpty()) return Optional.empty();
        OnePair second = secondOpt.get();

        return Optional.of(new TwoPair(first.cards, second.cards, second.kicker));
    }

    public static Optional<ThreeOfAKind> findThreeOfAKind(List<Card> cards) {
        Optional<List<Card>> tripleOpt = findSameCards(cards, 3);
        if (tripleOpt.isEmpty()) return Optional.empty();
        List<Card> triple = tripleOpt.get();

        List<Card> list = cards.stream()
                .filter(card -> card.strength() != triple.getFirst().strength())
                .toList();
        Optional<HighCard> kickerOpt = findHighCard(list);
        if (kickerOpt.isEmpty()) return Optional.empty();
        HighCard kicker = kickerOpt.get();

        return Optional.of(new ThreeOfAKind(new Triple<>(triple.get(0), triple.get(1), triple.get(2)), kicker.card));
    }

    public static Optional<Straight> findStraight(List<Card> cards) {
        List<Card> sortedCards = cards.stream()
                .sorted(Comparator.comparingInt(card -> card.strength().ordinal()))
                .toList();
        List<CardStrength> strengths = sortedCards.stream().map(Card::strength).toList();
        if (strengths.stream().distinct().count() != 5) return Optional.empty();
        if (strengths.get(4).ordinal() - strengths.get(0).ordinal() != 4) return Optional.empty();

        return Optional.of(new Straight(new Quintuple<>(
                sortedCards.get(0), sortedCards.get(1), sortedCards.get(2), sortedCards.get(3), sortedCards.get(4))));
    }

    public static Optional<Flush> findFlush(List<Card> cards) {
        if (cards.size() != 5) return Optional.empty();
        CardSuit suit = cards.getFirst().suit();
        List<Card> suited = cards.stream().filter(card -> card.suit() == suit).toList();
        if (suited.size() != 5) return Optional.empty();

        return Optional.of(
                new Flush(new Quintuple<>(suited.get(0), suited.get(1), suited.get(2), suited.get(3), suited.get(4))));
    }

    public static Optional<FullHouse> findFullHouse(List<Card> cards) {
        Optional<ThreeOfAKind> threeOfAKindOpt = findThreeOfAKind(cards);
        if (threeOfAKindOpt.isEmpty()) return Optional.empty();
        ThreeOfAKind threeOfAKind = threeOfAKindOpt.get();

        var groups = cards.stream().collect(java.util.stream.Collectors.groupingBy(Card::strength));
        var triple = groups.values().stream().filter(g -> g.size() == 3).findFirst();
        var pair = groups.values().stream().filter(g -> g.size() == 2).findFirst();

        if (triple.isEmpty() || pair.isEmpty()) return Optional.empty();

        var t = triple.get();
        var p = pair.get();

        return Optional.of(new FullHouse(new Triple<>(t.get(0), t.get(1), t.get(2)), new Pair<>(p.get(0), p.get(1))));
    }

    public static Optional<FourOfAKind> findFourOfAKind(List<Card> cards) {
        Optional<List<Card>> fourOfAKindOpt = findSameCards(cards, 4);
        if (fourOfAKindOpt.isEmpty()) return Optional.empty();
        List<Card> fourOfAKind = fourOfAKindOpt.get();

        List<Card> list = cards.stream()
                .filter(card -> card.strength() != fourOfAKind.getFirst().strength())
                .toList();
        Optional<HighCard> kickerOpt = findHighCard(list);
        if (kickerOpt.isEmpty()) return Optional.empty();
        HighCard kicker = kickerOpt.get();

        return Optional.of(new FourOfAKind(
                new Quadruple<>(fourOfAKind.get(0), fourOfAKind.get(1), fourOfAKind.get(2), fourOfAKind.get(3)),
                kicker.card));
    }

    public static Optional<StraightFlush> findStraightFlush(List<Card> cards) {
        Optional<Straight> straightOpt = findStraight(cards);
        if (straightOpt.isEmpty()) return Optional.empty();
        Straight straight = straightOpt.get();

        Optional<Flush> flushOpt = findFlush(cards);
        if (flushOpt.isEmpty()) return Optional.empty();

        return Optional.of(new StraightFlush(straight.cards));
    }

    private static Optional<List<Card>> findSameCards(List<Card> cards, int howMany) {
        var groups = cards.stream()
                .sorted(Comparator.comparingInt(card -> card.strength().ordinal()))
                .collect(java.util.stream.Collectors.groupingBy(Card::strength));

        return groups.values().stream().filter(g -> g.size() == howMany).findFirst();
    }
}

File: simple_poker_hands_kotlin/src/main/kotlin/Solver.kt
import kotlin.Result.Companion.failure
import kotlin.Result.Companion.success
import kotlin.collections.List

class Solver(
    private val cards: List<Card>,
) {
    fun solve(): Result<PokerHand> {
        val pokerHand =
            listOf(
                cards.findStraightFlush(),
                cards.findFourOfAKind(),
                cards.findFullHouse(),
                cards.findFlush(),
                cards.findStraight(),
                cards.findThreeOfAKind(),
                cards.findTwoPair(),
                cards.findOnePair(),
                cards.findHighCard(),
            ).firstNotNullOfOrNull { it.getOrNull() }

        if (pokerHand == null) {
            return failure(Exception("Not solved"))
        }

        return success(pokerHand)
    }
}

fun List<Card>.findHighCard(): Result<PokerHand.HighCard> {
    val card = this.maxByOrNull { it.strength } ?: return failure(Exception("Not HighCard"))

    return success(PokerHand.HighCard(card))
}

fun List<Card>.findOnePair(): Result<PokerHand.OnePair> {
    val pair =
        this
            .findSameCards(2)
            .getOrNull()
            ?: return failure(Exception("Not OnePair"))
    val kicker =
        this
            .filter { it.strength != pair.first().strength }
            .findHighCard()
            .getOrNull()
            ?: return failure(Exception("Not OnePair"))

    return success(
        PokerHand.OnePair(
            cards = Pair(pair[0], pair[1]),
            kicker = kicker.card,
        ),
    )
}

fun List<Card>.findTwoPair(): Result<PokerHand.TwoPair> {
    val firstPair =
        this
            .findOnePair()
            .getOrNull()
            ?: return failure(Exception("Not TwoPair"))
    val secondPair =
        this
            .filter { it.strength != firstPair.cards.first.strength }
            .findOnePair()
            .getOrNull()
            ?: return failure(Exception("Not TwoPair"))

    return success(
        PokerHand.TwoPair(
            first = firstPair.cards,
            second = secondPair.cards,
            kicker = secondPair.kicker,
        ),
    )
}

fun List<Card>.findThreeOfAKind(): Result<PokerHand.ThreeOfAKind> {
    val threeOfAKind =
        this
            .findSameCards(3)
            .getOrNull()
            ?: return failure(Exception("Not ThreeOfAKind"))
    val kicker =
        this
            .filter { it.strength != threeOfAKind.first().strength }
            .findHighCard()
            .getOrNull()
            ?: return failure(Exception("Not ThreeOfAKind"))

    return success(
        PokerHand.ThreeOfAKind(
            cards = Triple(threeOfAKind[0], threeOfAKind[1], threeOfAKind[2]),
            kicker = kicker.card,
        ),
    )
}

fun List<Card>.findStraight(): Result<PokerHand.Straight> {
    val sortedCards = this.sortedBy { it.strength }
    val strengths = sortedCards.map { it.strength }.toSet()
    if (strengths.size != 5) return failure(Exception("Not Straight"))
    if (strengths.last().ordinal - strengths.first().ordinal != 4) return failure(Exception("Not Straight"))

    return success(
        PokerHand.Straight(
            cards =
                Quintuple(
                    sortedCards[0],
                    sortedCards[1],
                    sortedCards[2],
                    sortedCards[3],
                    sortedCards[4],
                ),
        ),
    )
}

fun List<Card>.findFlush(): Result<PokerHand.Flush> {
    if (this.size != 5) return failure(Exception("Not Flush"))
    val suit = this.getOrNull(0)?.suit ?: return failure(Exception("Not Flush"))
    val suitedCards = this.filter { it.suit == suit }
    if (suitedCards.size != 5) return failure(Exception("Not Flush"))

    return success(
        PokerHand.Flush(
            Quintuple(
                suitedCards[0],
                suitedCards[1],
                suitedCards[2],
                suitedCards[3],
                suitedCards[4],
            ),
        ),
    )
}

fun List<Card>.findFullHouse(): Result<PokerHand.FullHouse> {
    val counts = this.groupBy { it.strength }
    val triples =
        counts.filter { it.value.size == 3 }.values.firstOrNull() ?: return failure(Exception("Not FullHouse"))
    val pairs = counts.filter { it.value.size == 2 }.values.firstOrNull() ?: return failure(Exception("Not FullHouse"))

    return success(
        PokerHand.FullHouse(
            first = Triple(triples[0], triples[1], triples[2]),
            second = Pair(pairs[0], pairs[1]),
        ),
    )
}

fun List<Card>.findFourOfAKind(): Result<PokerHand.FourOfAKind> {
    val fourOfAKind = this.findSameCards(4).getOrNull() ?: return failure(Exception("Not FourOfAKind"))
    val kicker =
        this.filter { it.strength != fourOfAKind[0].strength }.findHighCard().getOrNull()
            ?: return failure(Exception("Not FourOfAKind"))

    return success(
        PokerHand.FourOfAKind(
            cards = Quadruple(fourOfAKind[0], fourOfAKind[1], fourOfAKind[2], fourOfAKind[3]),
            kicker = kicker.card,
        ),
    )
}

fun List<Card>.findStraightFlush(): Result<PokerHand.StraightFlush> {
    val straight = this.findStraight().getOrNull() ?: return failure(Exception("Not StraightFlush"))
    this.findFlush().getOrNull() ?: return failure(Exception("Not StraightFlush"))

    return success(
        PokerHand.StraightFlush(
            cards = straight.cards,
        ),
    )
}

private fun List<Card>.findSameCards(howMany: Int): Result<List<Card>> {
    val groups = this.groupBy { it.strength }
    val matchingStrengths = groups.filter { it.value.size == howMany }.keys.sortedDescending()

    if (matchingStrengths.isEmpty()) {
        return failure(Exception("Not found $howMany same cards"))
    }

    return success(groups[matchingStrengths.first()]!!)
}
